### ISSUE TITLE
Accessing request.params breaks parameter encoding

### ISSUE THREAD

**Issue by fxn** [rails/rails#42278]
Given a controller:

```ruby
class FooController < ApplicationController
  param_encoding :index, :foo, Encoding::ASCII_8BIT
  skip_before_action :verify_authenticity_token

  def index
    head :ok
  end
end
```

The `index` parameter should accept any binary payload. For example, this should work

```ruby
require "net/http"

uri = URI("http://localhost:3000")
Net::HTTP.post_form(uri, "foo" => "\xe1") # Invalid UTF-8.
```

and it does.

However, a middleware as innocent as

```ruby
class FooMiddleware
  def initialize(app)
    @app = app
  end

  def call(env)
    request = ActionDispatch::Request.new(env)
    request.params
    @app.call(request.env)
  end
end
```

interferes somehow with parameter encoding validation, and the controller now raises

```
ActionController::BadRequest (Invalid request parameters: Invalid encoding for parameter: �)
```

**Comment by fxn** [comment]
Some progress.

The mere evaluation of `request.params` in the middleware, [triggers](https://github.com/rails/rails/blob/168a277bc63f5d40bbb21df1b19fa8f425f992fb/actionpack/lib/action_dispatch/http/parameters.rb#L55) the invocation of `AD::Request#request_parameters`, which [performs encoding validation](https://github.com/rails/rails/blob/168a277bc63f5d40bbb21df1b19fa8f425f992fb/actionpack/lib/action_dispatch/http/request.rb#L397).

However, since we have not reached the router yet, that code does not even have the information needed to do encoding validation. For example, you do not know the controller. The exception is raised by the middleware itself.

/cc @jhawthorn

**Comment by jhawthorn** [comment]
Possibly this changed in https://github.com/rails/rails/pull/40124 ?

**Comment by fxn** [comment]
I have not bisected, but that patch seems like a good candidate.

My thoughts:

1. Instantiating `AD::Request` in middleware to access and/or modify `env` is normal.
2. In particular, users should be able to access `request.params`, and the implementation which its cache invalidation, etc. suggest to me this is expected.

I believe we agree on that, no?

So, since parameter encoding validation depends on controller/action, from a logical point of view it does not make sense that we perform any parameter encoding validation before reaching the router.

Maybe that means the router is the one that has to validate the parameters somehow, cannot be embedded/automated in the request object like it is now.

Not sure if we need to guarantee validation above the router, if not, maybe the fix is moving things down. If yes, the design is not good, you just lack the necessary information.

Let's cc @adrianna-chang-shopify and @rafaelfranca.

**Comment by adrianna-chang-shopify** [comment]
Hey @fxn !

Can confirm that https://github.com/rails/rails/pull/40124 is likely the candidate (although validating the encoding on the request object was first introduced [here](https://github.com/rails/rails/pull/21990) to catch invalid querystring values, with https://github.com/rails/rails/pull/40124 simply extending that to `POST` requests).

Accessing `request.params` in a piece of middleware should certainly be doable, so it makes sense on my end to push this down to the router. Maybe `AD::Journey::Router#serve` could handle all of the encoding / validations ? I don't see a reason to need to validate encodings prior to hitting the router, but maybe @rafaelfranca has some extra context on that 😄

**Comment by fxn** [comment]
Perfect @adrianna-chang-shopify! 

Another option could be somewhere near [`ActionDispatch::Routing::RouteSet#call`](https://github.com/rails/rails/blob/328113b0adfb516639ea0ad9845e15877b8c8141/actionpack/lib/action_dispatch/routing/route_set.rb#L838-L842), since that seems to be the [lowest boundary](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/routing/route_set.rb#L32) before passing the request to the controller (if my limited understanding of this part of the framework is right).

Let's see what @rafaelfranca says!

**Comment by fxn** [comment]
Nah, can't be in `AD::Routing::RouteSet#call`, controller and action are set indeed in `AD::Journey::Router#serve`. Yeah, your suggestion is probably the best one.

**Comment by adrianna-chang-shopify** [comment]
Great! I should have some time to fix this later in the week, unless you'd like to tackle it @fxn ?

**Comment by fxn** [comment]
@adrianna-chang-shopify fantastic! all yours!

**Comment by rafaelfranca** [comment]
@fxn which character encoding do you expect `request.params` to use when you instantiate a `ActionDispatch::Request`?

If the controller is not know, the character encoding should be binary and I think the encoding validation would pass for any input.

I think that is a better change than making the parameters UTF-8 and not validating the input.

**Comment by fxn** [comment]
@rafaelfranca That's a great question.

I also believe it should be binary, meaning, we don't know at this point.

Because `AD::Request` is not a generic class that you can instantiate anywhere (for which [default encodings](https://github.com/rails/rails/blob/5adb6a9a98a33c699b184fdb6f38ddea740fbe86/railties/lib/rails.rb#L20-L24) would probably be the natural choice), but a class intended to help in middleware, and that models the request object in controllers. In this constrained context, I think the natural encoding is binary, and we can document that.

**Comment by fxn** [comment]
To give you a concrete example: This issue originated in middleware stack above a SendGrid webhook.

SendGrid sends email metadata, and params that tell you their encoding. So, you truly can't assume anything before reaching the controller.

**Comment by adrianna-chang-shopify** [comment]
@rafaelfranca and I had a chat about this and we've decided to move the [encoding into the controller](https://github.com/rails/rails/blob/main/actionpack/lib/action_controller/metal.rb#L200), with `AD::Request` using binary encoding. I'll update the PR I opened to move the encoding from the router into the controller and ensure we've documented the fact that encoding with `AD::Request` is binary since the controller context might not be available yet.

**Comment by fxn** [comment]
Fantastic @adrianna-chang-shopify!

**Comment by raldred** [comment]
Happy to test this, we too are seeing issue with processing binary params from Sendgrid's inbound parse API.

**Comment by adrianna-chang-shopify** [comment]
Finally had some time to dig into this a bit more, and if I'm not mistaken ensuring that params accessible in middleware are binary encoded will require an upstream change to Rack. Wanted to check in with some smart folks to make sure this all seems sensible before proceeding 😄 

For POST requests, we parse the query parameters [here](https://github.com/rack/rack/blob/d0c6efc666ede26768f33935f00530629690369a/lib/rack/request.rb#L448), and for GET requests we parse them [here](https://github.com/rack/rack/blob/master/lib/rack/request.rb#L424). Both end up calling [`parse_query`](https://github.com/rack/rack/blob/d0c6efc666ede26768f33935f00530629690369a/lib/rack/request.rb#L579-L581), which calls `query_parser.parse_nested_query`:

```ruby
# lib/rack/query_parser.rb
    def parse_nested_query(qs, separator = nil)
      params = make_params

      unless qs.nil? || qs.empty?
        (qs || '').split(separator ? (COMMON_SEP[separator] || /[#{separator}] */n) : DEFAULT_SEP).each do |p|
          k, v = p.split('=', 2).map! { |s| unescape(s) } # THIS LINE IS IMPORTANT

          normalize_params(params, k, v, param_depth_limit)
        end
      end

      return params.to_h
    rescue ArgumentError => e
      raise InvalidParameterError, e.message, e.backtrace
    end
```

When splitting the query string we use [`unescape`](https://github.com/rack/rack/blob/d0c6efc666ede26768f33935f00530629690369a/lib/rack/utils.rb#L53), which forces a UTF-8 encoding, meaning that when we access params in `AD::Request`, they are already encoded as UTF-8.

My proposition:
- `Rack::QueryParser#parse_query` takes an [`unescaper` block](https://github.com/rack/rack/blob/d0c6efc666ede26768f33935f00530629690369a/lib/rack/query_parser.rb#L35), allowing a custom method to be used for unescaping the query string
- We could adapt `Rack::QueryParser#parse_nested_query` to do the same
- This would allow us to override `#parse_query` in `AD::Request` and call `#query_parser.parse_nested_query` with an `unescape` method that ensures the encoding is binary (ie. `Rack::Utils.unescape(s, Encoding::ASCII_8BIT)`)

Does this make sense? cc @jhawthorn @fxn @rafaelfranca

**Comment by rafaelfranca** [comment]
We could change Rack but that would require Rails to force people to upgrade rack and I'd like to avoid that. We don't need to change the encoding of the parameters,  but we should move the validation of the encoding to the controller. It will be up to the middleware to check if the parameters they have are using a valid encoding.

**Comment by KSH-code** [comment]
Hey, 
I'm here to meet the problem which in upgrade rails from 6.0 to 6.1.

however, our project uses custom validation for params. so I'd like to disable all of validation from rails.
but from upgraded to rails 6.1, when invalid params entered, the error is raised always.


If the validation logic is moved to router, how do we disable it?

and I'm curious about how can disable in rails6.1.

**Comment by raldred** [comment]
We disable param encoding on our controller that receives inbound posts from third party services that are not supplying correctly encoded params.

We then manually sanitize the encoding when handling the params.

You can skip encoding with:
```
  skip_parameter_encoding :create
```

**Comment by KSH-code** [comment]
@raldred 
Thanks for the share.
my question is a my fault.
the real situation what I want is to rescue BadRequest and then respond `:bad_request` and customizing error message for api document.
is it possible??

**Comment by abury** [comment]
So, what is the fix for this issue in August 2022? `skip_parameter_encoding` doesn't work on an API controller.

**Comment by rahul342** [comment]
@abury were you able to figure this out for API controller?

**Comment by abury** [comment]
@rahul342 
Phoar it's been a while since I looked at this, but I'm fairly sure we ended up having to inherit from a different subclass for the affected controller.

**Comment by raldred** [comment]
@rahul342 We resorted to inheriting from `ActionController::Base` and using `skip_parameter_encoding` even for endpoints that we'd much rather use `ActionController::API`

**Comment by rahul342** [comment]
@rafaelfranca @adrianna-chang-shopify Is there a way to skip forced encoding on inheriting from `ActionController::API`. Or is there another way to send binary data in POST requests for API services?

**Comment by raldred** [comment]
@rahul342 You could try including `ActionController::ParameterEncoding` in your controller.

**Comment by rahul342** [comment]
Yep, just read through the code of `BaseController`, that makes sense. Thank you!
